src/pool.rs

```diff
@@ -709,24 +715,66 @@ impl<Conn: Connection + Send + 'static> Pool<Conn> {
         res
     }
 
+    // Acquiring a claim has two phases:
+    //
+    // 1. A request is made to the pool, to navigate backends and find a viable
+    //    claim which is believed to be connected.
+    // 2. Once such a claim is identified, it's returned to this calling task,
+    //    where the "Connector::on_acquire" method is invoked.
+    //
+    // We perform "on_acquire" seperately from the claim acquisition to ensure
+    // that the call to "on_acquire" does not limit access other claim requests
+    // in the pool.
     async fn do_claim(&self, id: ClaimId) -> Result<claim::Handle<Conn>, Error> {
-        let (tx, rx) = oneshot::channel();
-
-        // TODO: The work of "on_acquire" could be done here? Would prevent the
-        // slot set from blocking, and we'd also be able to loop. Picking a new
-        // backend might also be the right call if "on_acquire" is failing?
-        //
-        // There is admittedly a question of "fairness" within a queue of
-        // claims; looping at this point would send us to the back of the queue
-        // if there were multiple callers.
+        loop {
+            let (tx, rx) = oneshot::channel();
+            self.tx
+                .send(Request::Claim { id, tx })
+                .await
+                .map_err(|_| Error::Terminated)?;
+            let mut claim = rx.await.map_err(|_| Error::Terminated)??;
+
+            // We execute the "Connector::on_acquire" work here.
+            //
+            // Notably, this can fail: If it does, we drop the claim,
+            // and try to acquire another one.
+            //
+            // TODO: There isn't much "fairness" in this error case - if
+            // "on_acquire" is failing, the caller will go to the back of
+            // the queue of requests. However, it's (probably) not their fault
+            // the the connection setup failed - it may be worth building
+            // some mechanism to put them back in the front of the queue.
+
+            let result = tokio::time::timeout(
+                self.policy.set_config.health_check_timeout,
+                self.backend_connector.on_acquire(&mut claim),
+            )
+            .await;
+
+            let Ok(result) = result else {
+                event!(Level::TRACE, "Timeout performing 'on_acquire' on claim");
+                #[cfg(feature = "probes")]
+                probes::claim__acquire__failed!(|| (
+                    self.name.as_str(),
+                    id.0,
+                    "Timeout performing 'on_acquire'"
+                ));
+                continue;
+            };
+            if let Err(err) = result {
+                event!(Level::TRACE, "Failed performing 'on_acquire' on claim");
```

Comment:
this should probably include the error:
```suggestion
                event!(Level::TRACE, error=%err,"Failed performing 'on_acquire' on claim");
```